Core/Configuration/JsonConfiguration.cs

```diff
@@ -0,0 +1,419 @@
+                {
+                    Name = instance.Key,
+                    Path = instance.Value.GameDir()
+                }).ToList();
+
+                SaveConfig();
+            }
+        }
+
+        public IEnumerable<string> GetAuthTokenHosts()
+        {
+            lock (_lock)
+            {
+                return config.AuthTokens.Select(token => token.Host);
+            }
+        }
+
+
+        public bool TryGetAuthToken (string host, out string token)
+        {
+            lock (_lock)
+            {
+                foreach (AuthToken t in config.AuthTokens)
+                {
+                    if (t.Host == host)
+                    {
+                        token = t.Token;
+                        return true;
+                    }
+                }
+
+                token = "";
+                return false;
+            }
+        }
+
+        public void SetAuthToken (string host, string token)
+        {
+            lock (_lock)
+            {
+                bool found = false;
+                foreach (AuthToken t in config.AuthTokens)
+                {
+                    if (t.Host == host)
+                    {
+                        found = true;
+                        t.Token = token;
+                    }
+                }
+
+                if (!found)
+                {
+                    config.AuthTokens.Add(new AuthToken
+                    {
+                        Host = host,
+                        Token = token
+                    });
+                }
+
+                SaveConfig();
+            }
+        }
+
+        // <summary>
+        // Save the JSON configuration file. Only call this while you own _lock.
+        // </summary>
+        private static void SaveConfig()
+        {
+            string json = JsonConvert.SerializeObject(config, Formatting.Indented);
+            File.WriteAllText(configFile, json);
+        }
+
+        // <summary>
+        // Load the JSON configuration file. This will replace the current state.
+        // Only call this while you own _lock.
+        //
+        // If the configuration file does not exist, this will create it and then
+        // try to populate it with values in the registry left from the old system.
+        // </summary>
+        private void LoadConfig()
+        {
+            try
+            {
+                string json = File.ReadAllText(configFile);
+                config = JsonConvert.DeserializeObject<Config>(json);
+
+                if (config == null)
+                {
+                    config = new Config();
+                }
+
+                if (config.KspInstances == null)
+                {
+                    config.KspInstances = new List<KspInstance>();
+
+                }
+
+                if (config.AuthTokens == null)
+                {
+                    config.AuthTokens = new List<AuthToken>();
```

Comment:
The `Config` class should initialize these `List`s itself rather than relying on external code to set them.

Reply:
AuthTokens and KspInstances can be explicitly set to null in the config file, so these checks are still necessary, even with the values being initialized by Config.